Support for platforms without atomics?

I originally opened this on the wrong repo: https://github.com/Rahix/shared-bus/issues/10.

```
$ cargo check --target thumbv6m-none-eabi
    Checking shared-bus-rtic v0.2.1 (https://github.com/ryan-summers/shared-bus-rtic?branch=feature/spi-full-duplex#fad95b37)
error[E0599]: no method named `compare_exchange` found for struct `core::sync::atomic::AtomicBool` in the current scope
  --> /Users/bwstitt/.cargo/git/checkouts/shared-bus-rtic-b510fcc9e8973e1f/fad95b3/src/lib.rs:73:14
   |
73 |             .compare_exchange(false, true, Ordering::SeqCst, Ordering::SeqCst)
   |              ^^^^^^^^^^^^^^^^ method not found in `core::sync::atomic::AtomicBool`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0599`.
error: could not compile `shared-bus-rtic`.

To learn more, run the command again with --verbose.
```

I'm not sure there's any way to "fix" this. I think it might just be better to say that `thumbv6m-none-eabi` and other platforms without atomics are not supported.

I'm going to use a platform with thumbv7em-none-eabihf instead of the feather_m0 that I was planning on using since atomics are going to be helpful for a few things.

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -32,3 +32,28 @@ jobs:
         with:
           command: build
           args: --release
+
+  thumbv6:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - name: Install Rust
+        uses: actions-rs/toolchain@v1
+        with:
+          toolchain: stable
+          target: thumbv6m-none-eabihf
+          override: true
+      - name: Check
+        uses: actions-rs/cargo@v1
+        with:
+          command: check
+          args: --verbose
+      - name: Build Debug
+        uses: actions-rs/cargo@v1
+        with:
+          command: build --features thumbv6
+      - name: Build Release
+        uses: actions-rs/cargo@v1
+        with:
+          command: build
+          args: --release --features thumbv6

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -51,9 +51,9 @@ jobs:
       - name: Build Debug
         uses: actions-rs/cargo@v1
         with:
-          command: build --features thumbv6
+          command: build --feature thumbv6
       - name: Build Release
         uses: actions-rs/cargo@v1
         with:
           command: build
-          args: --release --features thumbv6
+          args: --release --feature thumbv6

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -1,4 +1,4 @@
-name: Armv7 Build
+name: Arm Build
 
 on:
   push:
@@ -8,7 +8,7 @@ on:
 
 jobs:
 
-  compile:
+  thumbv7:
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v2
@@ -41,7 +41,7 @@ jobs:
         uses: actions-rs/toolchain@v1
         with:
           toolchain: stable
-          target: thumbv6m-none-eabihf
+          target: thumbv6m-none-eabi
           override: true
       - name: Check
         uses: actions-rs/cargo@v1

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -51,9 +51,10 @@ jobs:
       - name: Build Debug
         uses: actions-rs/cargo@v1
         with:
-          command: build --feature thumbv6
+          command: build
+          args: --features thumbv6
       - name: Build Release
         uses: actions-rs/cargo@v1
         with:
           command: build
-          args: --release --feature thumbv6
+          args: --release --features thumbv6

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -16,3 +16,7 @@ readme = "README.md"
 
 [dependencies]
 embedded-hal = "0.2.4"
+cortex-m = { version = "0.6", optional = true }
+
+[features]
+thumbv6 = ["cortex-m"]

--- a/README.md
+++ b/README.md
@@ -9,6 +9,15 @@ prevent one driver from interrupting another while they are using the same under
 
 This crate also provides convenience types for working with `shared-bus` RTIC resources.
 
+## Features
+
+This crate is compatible with thumbv6 architectures. To enable support for thumbv6
+devices, enable the `thumbv6` feature in your `Cargo.toml`:
+```
+[dependencies.shared-bus-rtic]
+features = ["thumbv6"]
+```
+
 ## Usage Example
 ```rust
 

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -66,8 +66,7 @@ impl<BUS> CommonBus<BUS> {
     }
 
     fn lock<R, F: FnOnce(&mut BUS) -> R>(&self, f: F) -> R {
-        self.busy
-            .compare_exchange(false, true, Ordering::SeqCst, Ordering::SeqCst)
+        atomic::compare_exchange(&self.busy, false, true, Ordering::SeqCst, Ordering::SeqCst)
             .expect("Bus conflict");
         let result = f(unsafe { &mut *self.bus.get() });
 
@@ -125,6 +124,46 @@ impl<BUS: spi::Write<u8>> spi::Write<u8> for &CommonBus<BUS> {
 
     fn write(&mut self, words: &[u8]) -> Result<(), Self::Error> {
         self.lock(|bus| bus.write(words))
+    }
+}
+
+#[cfg(feature = "thumbv6")]
+mod atomic {
+    use core::sync::atomic::{AtomicBool, Ordering};
+
+    #[inline(always)]
+    pub fn compare_exchange(
+        atomic: &AtomicBool,
+        current: bool,
+        new: bool,
+        _success: Ordering,
+        _failure: Ordering,
+    ) -> Result<bool, bool> {
+        cortex_m::interrupt::free(|_cs| {
+            let prev = atomic.load(Ordering::Acquire);
+            if prev == current {
+                atomic.store(new, Ordering::Release);
+                Ok(prev)
+            } else {
+                Err(false)
+            }
+        })
+    }
+}
+
+#[cfg(not(feature = "thumbv6"))]
+mod atomic {
+    use core::sync::atomic::{AtomicBool, Ordering};
+
+    #[inline(always)]
+    pub fn compare_exchange(
+        atomic: &AtomicBool,
+        current: bool,
+        new: bool,
+        success: Ordering,
+        failure: Ordering,
+    ) -> Result<bool, bool> {
+        atomic.compare_exchange(current, new, success, failure)
     }
 }
 

